mp_api/client/mprester.py

```diff
@@ -1477,3 +1477,148 @@ def query(*args, **kwargs):
+        self,
+        material_id: MPID | str | list[MPID | str],
+        normalization: Literal["atom", "formula_unit"] = "atom",
+    ) -> float | dict[str, float]:
+        """Obtain the cohesive energy of the structure(s) corresponding to one or many MPIDs.
+
+        Args:
+            material_id (MPID or str or [MPID | str]) : a single MPID or a list of many to compute
+                cohesive energies for.
+            normalization (str = "atom" (default) or "formula_unit") :
+                Whether to normalize the cohesive energy by the number of atoms (default)
+                or by the number of formula units.
+                Note that the current default is inconsistent with the legacy API.
+
+        Returns:
+            (float or dict[str,float]) : If only MPID was specified, returns the cohesive energy
+                in eV/atom for that material. If multiple MPIDs were specified, the cohesive
+                energies (in eV/atom) for each material are returned in a dict indexed by MPID.
+        """
+        if isinstance(material_id, MPID | str):
+            material_id = [material_id]
+
+        entry_preference = {
+            k: i for i, k in enumerate(["GGA", "GGA_U", "SCAN", "R2SCAN"])
+        }
+        run_type_to_dfa = {"GGA": "PBE", "GGA_U": "PBE", "R2SCAN": "r2SCAN"}
+
+        energies = {mp_id: {} for mp_id in material_id}
+        entries = self.get_entry_by_material_id(material_id, compatible_only=False)
+        for entry in entries:
+            # Ensure that this works with monty_decode = False and True
+            if isinstance(entry, dict):
+                entry["uncorrected_energy_per_atom"] = entry["energy"] / sum(
+                    entry["composition"].values()
+                )
+            else:
+                entry = {
+                    "data": entry.data,
+                    "uncorrected_energy_per_atom": entry.uncorrected_energy_per_atom,
+                    "composition": entry.composition,
+                }
+
+            mp_id = entry["data"]["material_id"]
+            if (run_type := entry["data"]["run_type"]) not in energies[mp_id]:
+                energies[mp_id][run_type] = {
+                    "total_energy_per_atom": float("inf"),
+                    "composition": None,
+                }
+
+            if (
+                entry["uncorrected_energy_per_atom"]
+                < energies[mp_id][run_type]["total_energy_per_atom"]
+            ):
+                energies[mp_id][run_type] = {
+                    "total_energy_per_atom": entry["uncorrected_energy_per_atom"],
+                    "composition": entry["composition"],
+                }
+
+        atomic_energies = self.get_atom_reference_data()
+
+        e_coh_per_atom = {}
+        for mp_id, entries in energies.items():
+            if len(entries) == 0:
+                continue
+            prefered_func = sorted(list(entries), key=lambda k: entry_preference[k])[-1]
```

Comment:
probably doesn't make a difference but could also use `reverse=True` and index 0

Reply:
Added a comment as to why this indexing is used, there's a tiny performance penalty for reverse sorting